### PR TITLE
Improve LinearAlgebra impl methods for triangular types

### PR DESCRIPTION
**Pull request by maleadt**
Supersedes https://github.com/JuliaGPU/CUDA.jl/pull/550

### CODE REVIEW

**Comment by maleadt** [comment]
@amontoison Could you have a look if the CUSPARSE changes I ported from your PR look OK?

File: lib/cusparse/interfaces.jl [hunk 1]
```diff
@@ -71,66 +71,57 @@ for (taga, untaga) in tag_wrappers, (wrapa, transa, unwrapa) in op_wrappers
...
+        Base.:(\)(A::$t{T,<:AbstractCuSparseMatrix},
+                  B::DenseCuVector{T}) where {T<:BlasFloat} =
+            sv2('N', $uploc, $isunitc, one(T), parent(A), B, 'O')
+
+        # Matrix
+        Base.:(\)(A::$t{T,<:AbstractCuSparseMatrix},
+                  B::DenseCuMatrix{T}) where {T<:BlasFloat} =
+            sm2('N', 'N', $uploc, $isunitc, one(T), parent(A), B, 'O')
+    end
+end
```
Comment (amontoison) [inline lib/cusparse/interfaces.jl]:
```
# Left division
LinearAlgebra.ldiv!(A::$t{T,<:AbstractCuSparseMatrix},
                  B::DenseCuVector{T}) where {T<:BlasFloat} =
            sv2!('N', $uploc, $isunitc, one(T), parent(A), B, 'O')

LinearAlgebra.ldiv!(A::$t{T,<:AbstractCuSparseMatrix},
                  B::DenseCuMatrix{T}) where {T<:BlasFloat} =
            sm2!('N', 'N', $uploc, $isunitc, one(T), parent(A), B, 'O')
```

File: lib/cusparse/interfaces.jl [hunk 2]
```diff
@@ -71,66 +71,57 @@ for (taga, untaga) in tag_wrappers, (wrapa, transa, unwrapa) in op_wrappers
...
+        # Vector
+        Base.:(\)(A::$t{<:Any,<:Transpose{T,<:AbstractCuSparseMatrix}},
+                  B::DenseCuVector{T}) where {T<:BlasFloat} =
+            sv2('T', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')
+        Base.:(\)(A::$t{<:Any,<:Adjoint{T,<:AbstractCuSparseMatrix}},
+                  B::DenseCuVector{T}) where {T<:BlasReal} =
+            sv2('T', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')
+        Base.:(\)(A::$t{<:Any,<:Adjoint{T,<:AbstractCuSparseMatrix}},
+                  B::DenseCuVector{T}) where {T<:BlasComplex} =
+            sv2('C', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')
```
Comment (amontoison) [inline lib/cusparse/interfaces.jl]:
```
# Left division
LinearAlgebra.ldiv!(A::$t{<:Any,<:Transpose{T,<:AbstractCuSparseMatrix}},
                  B::DenseCuVector{T}) where {T<:BlasFloat} =
            sv2!('T', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')

LinearAlgebra.ldiv!(A::$t{<:Any,<:Adjoint{T,<:AbstractCuSparseMatrix}},
                  B::DenseCuVector{T}) where {T<:BlasReal} =
            sv2!('T', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')

LinearAlgebra.ldiv!(A::$t{<:Any,<:Adjoint{T,<:AbstractCuSparseMatrix}},
                  B::DenseCuVector{T}) where {T<:BlasComplex} =
            sv2!('C', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')
```

File: lib/cusparse/interfaces.jl [hunk 3]
```diff
@@ -71,66 +71,57 @@ for (taga, untaga) in tag_wrappers, (wrapa, transa, unwrapa) in op_wrappers
...
+        # Matrix
+        Base.:(\)(A::$t{<:Any,<:Transpose{T,<:AbstractCuSparseMatrix}},
+                  B::DenseCuMatrix{T}) where {T<:BlasFloat} =
+            sm2('T', 'N', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')
+        Base.:(\)(A::$t{<:Any,<:Adjoint{T,<:AbstractCuSparseMatrix}},
+                  B::DenseCuMatrix{T}) where {T<:BlasReal} =
+            sm2('T', 'N', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')
+        Base.:(\)(A::$t{<:Any,<:Adjoint{T,<:AbstractCuSparseMatrix}},
+                  B::DenseCuMatrix{T}) where {T<:BlasComplex} =
+            sm2('C', 'N', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')
```
Comment (amontoison) [inline lib/cusparse/interfaces.jl]:
```
# Left division
LinearAlgebra.ldiv!(A::$t{<:Any,<:Transpose{T,<:AbstractCuSparseMatrix}},
                  B::DenseCuMatrix{T}) where {T<:BlasFloat} =
            sm2!('T', 'N', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')

LinearAlgebra.ldiv!(A::$t{<:Any,<:Adjoint{T,<:AbstractCuSparseMatrix}},
                  B::DenseCuMatrix{T}) where {T<:BlasReal} =
            sm2!('T', 'N', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')

LinearAlgebra.ldiv!(A::$t{<:Any,<:Adjoint{T,<:AbstractCuSparseMatrix}},
                  B::DenseCuMatrix{T}) where {T<:BlasComplex} =
            sm2!('C', 'N', $uploc, $isunitc, one(T), parent(parent(A)), B, 'O')
```

**Review by amontoison — CHANGES_REQUESTED** [review]
I had a look and everything was fine for me.
I just found missing `ldiv!` methods that we can easily add.

**Comment by maleadt** [comment]
Looks like we can just get rid of the out-of-place methods and only provide `ldiv!`, because of https://github.com/JuliaLang/julia/blob/bdcaee033e8913459b340a2116e667770b4f0bc7/stdlib/LinearAlgebra/src/triangular.jl#L1650-L1691 ?

**Comment by amontoison** [comment]
Yes, well spotted! We can only provide `ldiv!`.

**Pull request merged**